stdlib/functools.pyi

```diff
@@ -51,22 +53,46 @@ if sys.version_info >= (3, 9):
         maxsize: int
         typed: bool
 
+class _Function(Protocol[_P, _R]):  # type: ignore[misc]  # pyright: ignore reportInvalidTypeVarUse
+    def __call__(self, *args: _P.args, **kwds: _P.kwargs) -> _R: ...
+
+class _Method(Protocol[_T, _P, _R]):  # type: ignore[misc]  # pyright: ignore reportInvalidTypeVarUse
+    def __call__(_self, self: _T, *args: _P.args, **kwds: _P.kwargs) -> _R: ...
+
+class _Classmethod(Protocol[_T, _P, _R]):  # type: ignore[misc]  # pyright: ignore reportInvalidTypeVarUse
+    def __call__(_self, cls: type[_T], *args: _P.args, **kwds: _P.kwargs) -> _R: ...
```

Comment:
I think combine these and use `Concatenate` instead, avoid needing to explicitly name a parameter, fix issue mentioned by @Daverball

Reply:
I think that's pretty much what the closest to working earlier approach did, yes. But it still had its edge cases where it didn't work unfortunately. Maybe things are a little bit more forgiving now.

If I'm not mistaken what you're doing in addition to earlier approaches is to always accept both signatures for methods/classmethods (i.e. the one where you have to explicitly provide `self`/`cls` and the one where you don't), so you give up some type safety on `instance.method()` vs. `Cls.method()` since it would not be able to detect that you're missing a required parameter.

That might be a reasonable trade-off for the overall better type safety, although I fear that language servers will not know what to do with these overloads, so instead of no parameter hints you may get confusing/difficult to read hints, which could be seen as a regression in ergonomics.